### ISSUE TITLE
ELC misclassified as electrostatic extension

### ISSUE THREAD

**Issue by fweik** [espressomd/espresso#2155]
In Espresso ELC is implemented as an electrostatics solver, and should be interfaced as such. This has the effect that it can not be combined with ICC, which should be possible.

**Comment by jngrad** [comment]
Another issue of the current implementation is that we cannot remove an ELC actor from the list of active actors, because it raises an exception (this is a workaround for dependency inversion reasons). Setting up the ELC actor sets the ELC core struct but also alters the P3M core struct, and this change cannot be reverted.

Using composition, we can have the ELC actor take a P3M actor as argument of the ELC constructor. Then only the ELC actor is added to the list of active actors, i.e.:

```python
p3m = electrostatics.P3M(prefactor=1.0, accuracy=1e-2)
elc = electrostatics.ELC(p3m_actor=p3m, maxPWerror=1e-1, gap_size=elc_gap)
system.actors.add(elc)
```

This solution works out-of-the-box, including checkpointing.

The alternative is inheritance, i.e. having ELC as a child class of P3M. This is not straightforward:
* one cannot write `cdef class ELC(P3M)` because the `P3M` class is overshadowed by the `P3M` macro; one also cannot use the usual workaround `getattr(sys.modules[__name__], "P3M")` to get the class type instead of the macro, because in python the parent class must be written as a symbol, not an expression. Here the solution is to rename the `P3M` class to `P3MCPU`, create an empty class `cdef class P3M(P3MCPU)` for backward compatibility, and then `cdef class ELC(P3MCPU)`
* the ELC changes the P3M struct, and it's unclear to me what happens when you call `tune()` on that modified P3M struct
* when setting the parameters in the core, first the P3M parameters then the ELC parameters, the `gap_size` gets reset to the last value initialized in the core (0 if this is the first ELC actor) because the P3M tuning function calls `_get_params_from_es_core()` before the ELC parameters were written to the core, so we read whatever what was left at the memory address of `elc_params` and update the shared ELC/P3M `_params` dict
* I'm not sure what will happen when loading from a checkpoint

I don't think having ELC as a child class of P3M is a good strategy, because it doesn't reflect how ELC is currently implemented in the core. The core structs for P3M and ELC were designed in such a way that the ELC state is set up after P3M is in a valid and tuned state. The composition solution works because it acts like a context manager, and the python `_params` dict member of the Actor classes is not shared between the 2 actors.

**Comment by jngrad** [comment]
Depends on #4124 and #4121.
